Only act on confirmed spends

halseth opened the pull request:
This PR removes the option to register spend notifications that will be triggered on spends detected in the mempool. This is done to ensure unified behavior across all backends, as they don't necessarily (most notably Neutrino) have a mempool.

This is a step in enabling integration tests also with `bitcoind` and `neutrino` backends.

### Structure
The commits are linearly ordered in such a way that one–by–one spend registration is changed over from a mempool spend to a confirmed spend, accompanied by any test changes that are needed to make them pass. This means that any commit that doesn't have a corresponding test change, **_should_** pass the tests without additional changes. 

Finally we remove the option for mempool spends altogether, when all usages are eliminated.

### Integration test changes
The most notable test change is modifications to `lnd_test`, which must be modified to account for the nodes not reacting to spends before they are confirmed. Additional comments and checks are added to make sure this works as expected.

### TODO
- [x] Why are we mining `CSV_DELAY + 1` blocks to trigger act on force closes?
- [x] Must be rebased on #1554

Roasbeef commented on contractcourt/contract_resolvers.go:
```diff
@@ -169,45 +169,17 @@ func (h *htlcTimeoutResolver) Resolve() (ContractResolver, error) {
-			return fmt.Errorf("quitting")
-		}
-
-		// Now that the output has been spent, we'll also wait for the
```
We still may want to actually wait for an additional, or set of additional confirmations...

halseth replied:
Note that this is equivalent to the previous logic:

Before this commit we wait for a spend (any spend, even mempool), then check that it has at least 1 conf.
After this commit we wait for a confirmed spend.

Are you suggestion that we should wait for additional confirmations of the spend? Should we do 6 conf?

Roasbeef replied:
Yes suggesting that we may want to wait for additional confirmations. This isn't a blocker for this PR though, as it would be a project wide change if we went down this route.

Roasbeef commented on contractcourt/chain_arbitrator.go:
```diff
@@ -514,13 +514,16 @@ func (c *ChainArbitrator) Stop() error {
 func (c *ChainArbitrator) watchForChannelClose(closeInfo *channeldb.ChannelCloseSummary) {
 	spendNtfn, err := c.cfg.Notifier.RegisterSpendNtfn(
-		&closeInfo.ChanPoint, closeInfo.CloseHeight, true,
+		&closeInfo.ChanPoint, closeInfo.CloseHeight, false,
```
Similar comment here w.r.t also enabling callers to wait on an additional set of confirmations.

Roasbeef commented on lnd_test.go:
```diff
@@ -5551,40 +5551,65 @@ func testRevokedCloseRetributionRemoteHodl(net *lntest.NetworkHarness,
 	// broadcast as Carol's contract breaching transaction gets confirmed
 	// above. Since Carol might have had the time to take some of the HTLC
-	// outputs to the second level before Alice broadcasts her justice tx,
+	// outputs to the second level before Dave broadcasts his justice tx,
```
Broadcast failure? Why's it possible for this to happen? Is it due to the new requirement of ensue the second level spends are now confirmed?

halseth replied:
Yes, since Dave would previously act directly on the second level spend in the mempool, this would eventually resolve. Now, with the removal of mempool spend checks, we need to mine an additional block for Dave to get notified of the second level tx.

Roasbeef commented on lnd_test.go:
```diff
@@ -5551,40 +5551,65 @@ func testRevokedCloseRetributionRemoteHodl(net *lntest.NetworkHarness,
+		return true
+	}, time.Second*10)
+	if err != nil && predErr == errNotFound {
+		fmt.Println("not found, mine new block")
```
Stray print statement.

halseth commented:
Blocked by #1554

halseth commented:
Rebased.

halseth commented:
Rebased! 💪 

For the record: we discussed the reason for mining +1 blocks during tests, and this is because the UTXO nursery broadcasts the transaction one block after it is actually first valid. We decided to keep this as is, since it doesn't really matter and many of the tests are written with this in mind. I modified my tests to match the patter: mine a block that includes the CSV locked tx, then mine `CSV_DELAY` number of blocks before sweep.

halseth commented:
Travis started failing after rebase on master. Investigating...

halseth commented:
We green 💚
